qutip/hsolve.py

```diff
@@ -0,0 +1,310 @@
+    Returns
+    -------
+    state_number : list
+        Successive state number arrays that can be used in loops and other
+        iterations, using standard state enumeration *by definition*.
+    """
+
+    if state is None:
+        state = np.zeros(len(dims))
+
+    if excitations and sum(state[0:idx]) > excitations:
+        pass
+    elif idx == len(dims):
+        if excitations is None:
+            yield np.array(state)
+        else:
+            yield tuple(state)
+            
+    else:
+        for n in range(dims[idx]):
+            state[idx] = n
+            for s in state_number_enumerate(dims, excitations, state, idx + 1):
+                yield s
+
+
+def cot(x):
+    """
+    Calculate cotangent.
+    Parameters
+    ----------
+    x: Float
+        Angle.
+    """
+    return np.cos(x)/np.sin(x)
+
+
+def hsolve(H, psi0, tlist, Q, gam, lam0, Nc, N, w_th, options=None):
+    """
+    Function to solve for an open quantum system using the
+    hierarchy model. 
+
+    Parameters
+    ----------
+    H: Qobj
+        The system hamiltonian.
+    psi0: Qobj
+        Initial state of the system.
+    tlist: List.
+        Time over which system evolves.
+    Q: Qobj
+        The coupling between system and bath.
+    gam: Float
+        Bath cutoff frequency.
+    lam0: Float
+        Coupling strength.
+    Nc: Integer
+        Cutoff parameter.
+    N: Integer
+        Number of matsubara terms.
+    w_th: Float
+        Temperature. 
+    options : :class:`qutip.Options`
+        With options for the solver.
+     
+    Returns
+    -------
+    output: Result
+        System evolution.
+    """
+    if options is None:
+        options = Options()
+    
+    # Set up terms of the matsubara and tanimura boundaries
+
+    # Parameters and hamiltonian
+    hbar=1.
+    kb=1.
+
+    # Set by system 
+    dimensions = dims(H)
+    Nsup = dimensions[0][0] * dimensions[0][0] 
+    unit = qeye(dimensions[0])
+
+    # Ntot is the total number of ancillary elements in the hierarchy
+    Ntot = int(round(factorial(Nc+N) / (factorial(Nc) * factorial(N))))
+    c0 =(lam0 * gam * (cot(gam * hbar / (2. * kb * w_th)) - (1j))) / hbar
+    LD1 = -2.* spre(Q) * spost(Q.dag()) + spre(Q.dag() * Q) + spost(Q.dag() * Q)
+    pref = ((2. * lam0 * kb * w_th / (gam * hbar)) -1j * lam0) / hbar
+    gj = 2 * np.pi * kb * w_th / hbar
+    L12 = -pref * LD1 + (c0 / gam) * LD1 
+
+    for i1 in range(1,N):
+        ci = (4 * lam0 * gam * kb * w_th * i1
+        	  * gj/((i1 * gj)**2 - gam**2)) / (hbar**2)
+        L12 = L12 + (ci / gj) * LD1
+
+    # Setup liouvillian
+
+    L = liouvillian(H, [L12])
+    Ltot = L.data
+    unit = sp.csr_matrix(np.identity(Ntot))
+    Lbig = sp.kron(unit,Ltot.tocsr())
+    rho0big1 = np.zeros((Nsup * Ntot), dtype=complex)
+
+
+    # Prepare initial state:
+
+    rhotemp =  mat2vec(np.array(psi0.full(), dtype=complex))
+
+    for idx,element in enumerate(rhotemp):
+        rho0big1[idx] = element[0]
```

Comment:
in the end we use the dense form of this for the initial state in the ODE solver (the sparse copy is never used so will be removed).  The ODE solver complains if I use a sparse form for the initial condition, is there anyway around that?